src/test/java/com/google/sps/ChatServletTest.java

```diff
@@ -0,0 +1,258 @@
+package com.google.sps;
+
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.sps.servlets.ChatServlet;
+import com.google.template.soy.SoyFileSet;
+import com.google.template.soy.tofu.SoyTofu;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Scanner;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public final class ChatServletTest extends Mockito {
+    private static final String GROUP_ID_PARAM = "groupID";
+    private static final String GROUP_ID = "123";
+    private static final String MESSAGE_TEXT_TOXIC = "what kind of idiot name is foo?";
+    private static final String MESSAGE_KIND = "Message-";
+    private static final String MESSAGE_TEXT_PROPERTY = "message-text";
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String GROUP_KIND = "Group";
+    private static final String GROUP_NAME_PROPERTY = "name";
+    private static final String MESSAGES_MAP_KEY = "messages";
+    private static final String GROUPS_KEY = "groups";
+    private static final ImmutableList<String> SAMPLE_GROUP_LIST = 
+        ImmutableList.of("Black Lives Matter");
+    private static final String CHAT_TEMPLATE = "templates.chat.chatPage";
+    private static final Double SCORE_OFFSET = 0.0000005;
```

Comment:
For the constants that are the same as the ones used in your classes, please export them as static variables used for testing. Variables shared amongst multiple classes should go into a constants file, and variables only shared between a class and its test should be exported as static variables.